docs: restructure working PR (#4866)
dfinity/motoko: doc/md/fundamentals/2-types/6-objects-classes.md

---
sidebar_position: 6
---

# Objects and classes

## Objects

> An object, similar to a [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records), stores structured data with optional mutable fields and supports methods, including [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) behavior. Unlike records, objects can encapsulate or share their [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior using `public` and `private` visibility modifiers. However, they cannot be instantiated independently, such as `object()`.

Reviewer: ```suggestion
In Motoko, an object is just a collection of named fields, holding values. These values can either be plain data, or function values. In addition, each field can be mutable or immutable.

A simple object containing just fields of data is like a record in a database.
Motoko's light-weight record syntax make it easy do construct such objects ([record](https://internetcomputer.org/docs/motoko/fundamentals/types/records).

When fields contain function values, Motoko objects can represent traditional objects with methods, familiar from object-oriented programming (OOP).
From an OOP perspective, an object is an abstraction, defined by the behavior of its methods. Methods are typically used to modify or observe some encapsulated (i.e. hidden) state of an object.

In addition to the record syntax, Motoko let's you defined objects from a block of declarations.
The declarations in the block can be `public` or `private`, with `private` the default.
The public declarations become fields of the object, all private declarations are hidden.
```

```motoko no-repl
object Account {
    var balance : Nat = 1000;

>     func deposit(amount : Nat) : Nat {

Reviewer: ```suggestion
   public  func deposit(amount : Nat) : Nat {
```

Reviewer: (otherwise this object has no fields)

        balance += amount;
        balance
    };

    func withdraw(amount : Nat) : ?Nat {
        if (amount > balance) { return null };
        balance -= amount;
        ?balance
    };
}
```

## Classes

A class acts as a blueprint for creating multiple objects with independent [state](https://internetcomputer.org/docs/motoko/fundamentals/state).

```motoko no-repl
class Account(initialBalance : Nat) {
    var balance = initialBalance;

    public func deposit(amount : Nat) : async Nat {
        balance += amount;
        balance
    };

    public func withdraw(amount : Nat) : ?Nat {
        if (amount > balance) { return null };
        balance -= amount;
        ?balance
    };
};

// Creating multiple accounts
let account1 = Account(500);
let account2 = Account(1000);
```

### Object classes

> An object class defines a blueprint for multiple objects.

Reviewer: The above is just short-hand for an `object` class, creating object.

> ### Object classes
> 
> An object class defines a blueprint for multiple objects.

Reviewer: ```suggestion
The above is just short-hand for an `object` class, a class that creates objects. 
```

```motoko no-repl
object class Account(initialBalance : Nat) {
    var balance = initialBalance;

    public func deposit(amount : Nat) : Nat {
        balance += amount;
        balance
    };

    public func withdraw(amount : Nat) : ?Nat {
        if (amount > balance) { return null };
        balance -= amount;
        ?balance
    };
}
```

> 

Reviewer: ```suggestion
Motoko also support module and actor classes.
```

## Modules

> A module provides reusable utility functions and encapsulates both [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior, module [state](https://internetcomputer.org/docs/motoko/fundamentals/state) is not persistent unless explicitly stored. It supports [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) operations, making it suitable for managing complex functionality across different parts of a program.

Reviewer: ```suggestion
Motoko also supports modules. Modules are similar to objects, containing public and private declarations, but are restricted to be stateless. They are typically used to implement libraries of types, functions and values, and, unlike objects, can be imported from other files.
```

Objects with fewer fields are more general, while objects with additional fields are subtypes of more general types.

| Generality         | Object type | Relation |
|--------------------|------------------------------------------|----------------------|
| Most general   | `object basicAccount = { getBalance : () -> Nat }` | Subtype of `standardAccount`  |
| Middle generality | `object standardAccount { getBalance : () -> Nat; deposit : Nat -> () }` | Subtype of `premiumAccount` |
| Least general  | `{ object premiumAccount getBalance : () -> Nat; deposit : Nat -> (); withdraw : Nat -> Bool }` | Supertype |

A function expecting `{ getBalance : () -> Nat }` can accept any of the above, since all contain at least that method. However, a function requiring `{ withdraw : Nat -> Bool }` cannot accept more general types that lack this method.

- A subtype must be usable wherever its supertype is expected.
- The more general object has fewer methods because it makes fewer assumptions about available functionality.

> `basicAccount <: standardAccount <: premiumAccount`, or equivalently:  `basicAccount` is a subtype of `standardAccount`, which is a subtype of `premiumAccount`.

Reviewer: ```suggestion
` premiumAccount <: standardAccount <: basicAccount`, or equivalently:  `premiumAccount` is a subtype of `standardAccount`, which is a subtype of `basicAccount`.
```

A function expecting `premiumAccount` expects `withdraw`, so it cannot accept `basicAccount`.

> However, a function expecting `basicAccount` only needs `getBalance`, so it can accept all three objects.

Reviewer: ```suggestion
However, a function expecting `basicAccount` only needs `getBalance`, so it can accept all three type of objects.
```

```motoko no-repl
func printBalance(account : { getBalance : () -> Nat }) {
  Debug.print("Balance: " # Nat.toText(account.getBalance()));
};

// Works, because all have getBalance
printBalance(basicAccount);
printBalance(standardAccount);
printBalance(premiumAccount);
```

```motoko no-repl
func withdrawFromAccount(account : { withdraw : Nat -> Bool }) {
  let success = account.withdraw(100);
  Debug.print(if success then "Withdrawal successful" else "Insufficient funds");
};

// Works only for premiumAccount, fails for others
withdrawFromAccount(premiumAccount);  // Works

> withdrawFromAccount(standardAccount); // Fails (missing withdraw)
> withdrawFromAccount(basicAccount);    // Fails (missing withdraw)

Reviewer: ```suggestion
withdrawFromAccount(standardAccount); // type error: (missing withdraw)
withdrawFromAccount(basicAccount);    // type error: (missing withdraw)
```
These don't fail dynamically but statically!
